introduce dynamism library

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147981

This is the first step in supporting delayed compile. This library takes in example inputs and outputs a dict of dynamism across the inputs. We will use this to detect dynamism across multiple inputs in delayed compile. We will also use this to make shape collections more ergonomic by providing an affordance to generate a shape collection using example inputs.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

bobrenjc93 committed:
introduce dynamism library

[ghstack-poisoned]

pianpwk commented on test/fx/test_dynamism.py:
```diff
@@ -0,0 +1,115 @@
+                "L['self']['_modules']['y']['x']": (True,),
+                "L['self']['n_input']": (True,),
+                "L['self']['n_output']": (False,),
+            }
```
Is it more practical to use this source mapping, rather than replicate the nested structure, export dynamic shapes spec style? Asking just because I don't know how this is used.

bobrenjc93 replied:
I personally think it'd be good to unify compile and export, and yes compile generally does source level dynamism annotations. Perhaps this will be a good forcing function to unify? You can see how this is used in the next PR in the stack.

bobrenjc93 commented:
@Pytorchbot merge

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 commented:
(for folks reading this from chat the first one was from mobile and second one from desktop)

wdvr commented:
@pytorchmergebot revert -m 'lint failure' -c landrace

bobrenjc93 commented:
@pytorchbot  merge

bobrenjc93 commented:
@pytorchbot test

malfet commented:
@pytorchmergebot revert -m 'lint failure' -c landrace

malfet reviewed (changes requested):
Please fix lint before trying to merge again

bobrenjc93 commented:
@malfet will do, I do find it odd that our land workflow doesn't check it. It makes sense that there might be a land race sometimes, but almost 14+ hours later? There is no lint signal on the actual PR so I assumed it was a land specific problem.

bobrenjc93 committed:
Update on "introduce dynamism library"


This is the first step in supporting delayed compile. This library takes in example inputs and outputs a dict of dynamism across the inputs. We will use this to detect dynamism across multiple inputs in delayed compile. We will also use this to make shape collections more ergonomic by providing an affordance to generate a shape collection using example inputs.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

wdvr commented:
@bobrenjc93 something seems wrong on the last commit:
![image](https://github.com/user-attachments/assets/f46d49d7-c926-4838-b06c-899799bb4f7d)

bobrenjc93 commented:
@wdvr hmm I actually can't find that commit. Can you share the url? Looking at the last commit and files changed from my end, things look normal.

wdvr commented:
https://github.com/pytorch/pytorch/commit/6ed77490ef50c3ddfe9171c096c6fc2cd1b632b9

wdvr commented:
Ah it's because it's a merge commit - on web, click the button 'You are viewing a condensed version of this merge commit. You can view the full changes here.'.

I guess on mobile GitHub defaults to merge commit

wdvr commented:
@pytorchmergebot merge -f "lint passed, not waiting for other tests "
